### DISCUSSION TITLE
Composable atom utils

### DISCUSSION THREAD

**Discussion by barucAlmaguer** [pmndrs/jotai#2587] [Ideas] [answered]
Hi!
I see that there are several utilities like AtomFamily, atomWithImmer, atomWithStorage, atomWithHash, etc.
But I think there are plenty of cases where it would be beneficial to have a way to compose functionalities.

For example, having both withImmer and withHash, to allow doing something like:

```
const myBaseAtom = atom(10)
const myImmerHashAtom = withHash(withImmer(myBaseAtom), {key: "limit", default: 5})

// then:
const setHashAtom = useSetAtom(myImmerHashAtom)
const onClick = () => {setHashAtom(draft => draft += 10)} // also changes url query param "limit" to 15
``` 

What do you think?

Also, is there documentation about how to create our own atom utilities?

**Comment by dai-shi** [comment, marked as answer]
Yes, withFoo would be nice. It's currently a limitation of the API. We'll investigate further to design some new API to allow such a pattern. It will take time!

**Comment by dai-shi** [comment]
Not a doc, but you might find this useful: https://blog.axlight.com/posts/jotai-tips/
